### PR TITLE
Initial mockup of memory stats postprocessor

### PR DESCRIPTION
**Pull request by spco**
Addresses #1052 

@tjhei would you be able to give some feedback on this very rough setup? Obviously lots more stats to add and tidying up of the description etc.

Outputs to both the statistics file and screen output. 

I can either add matrix stats a la #1053 in this, or add it to the same postprocessor/different postprocessor later.

### CODE REVIEW

File: source/postprocess/parallel_statistics.cc [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+namespace aspect
+{
+    namespace Postprocess
+    {
+        ASPECT_REGISTER_POSTPROCESSOR(ParallelStatistics,
+                                      "parallel statistics",
+                                      "A postprocessor that computes some statistics about "
+                                      "the parallel solution process. "
+                                      "In particular, it computes minimal, average and maximal "
+                                      "values of particles per cell in the global domain.")
```
Comment (bangerth) [inline source/postprocess/parallel_statistics.cc]: values of -> number of

But this doesn't seem to match what you do above. Is this a copy-paste issue?

**Comment by bangerth** [comment]
This could potentially produce many more columns in the statistics file. What about writing this kind of information into a separate file? (Not sure this is the right idea, but looking for a discussion.)

File: source/postprocess/parallel_statistics.cc (hunk 1 above)
Comment (spco) [inline source/postprocess/parallel_statistics.cc]: Sure - I have only half-rewritten this, as I though I'd write it last once we know what this function truly does!

**Comment by spco** [comment]
I agree that this could end up with a large number of columns, probably 10 or so depending on the desired granularity of the statistics. I guess we could have a second `TableHandler memory_statistics`, with `Simulator::output_memory_statistics()` function outputting to a `memory_statistics` file. I don't think it would be too much work, but it's whether that is a desirable solution - do we want 2 stats files? Are there other scenarios where we might want a third such file, for monitoring some other stats?

**Comment by bangerth** [comment]
I thought we can just keep the extra `TableHandler` inside the plugin itself. This way, it's just another, self-contained plugin.

**Comment by tjhei** [comment]
I think it depends on how many columns you plan on generating. If it is about 10 or so, I would say to just leave in ``statistics``. That makes plotting data much easier and doesn't do any harm. (I am excited that we can look at things like memory consumption over time or over number of DoFs!)

For something like memory stats for every single matrix block ( #1053) I would prefer a different solution. Of course there we probably don't care about statistics over time.

**Comment by spco** [comment]
I think having a local `TableHandler` would be better, for sure.

The other issue would be how to plot this data. Certainly the `memory_statistics` file should contain the time step number. But what about if I wanted to plot against something else, eg number of dofs (to name just one fairly likely case)? If `memory_statistics` doesn't contain this data, then in gnuplot you need to do something like 
`plot "< join statistics memory_statistics" using 1:27`

I'll implement it into `statistics`, then look later at the matrix usage.

**Comment by bangerth** [comment]
Ha, did not know about `join` :-)

My gut feeling is that over time, we will want to add *lots* of different statistics, and that we will be happier if those go into their own statistics file.

**Comment by spco** [comment]
What's here now is my 'finished' version. Let me know your thoughts.

File: source/postprocess/memory_statistics.cc [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+namespace aspect
+{
+  namespace Postprocess
+  {
+    ASPECT_REGISTER_POSTPROCESSOR(MemoryStatistics,
+                                  "memory statistics",
+                                  "A postprocessor that computes some statistics about "
+                                  "the memory consumption. "
+                                  "In particular, it computes the memory usage of the "
+                                  "system matrix, triangulation, (of which p4est), "
```
Comment (bangerth) [inline source/postprocess/memory_statistics.cc]: grammar

File: source/global.cc [hunk 3]
```diff
@@ -100,6 +100,4 @@ namespace aspect
     }
   }
 
-  // Output parallel statistics [bool]
-  const bool output_parallel_statistics = false;
```
Comment (bangerth) [inline source/global.cc]: is there an associated parameter that is being declared and read? If so, what should we do with it?

**Comment by bangerth** [comment]
I guess this is ok with me. We can always at a later point decide that we want to use a separate file for things.

Do you think you could write a test? Or do we think that this can not reasonably be tested?

File: source/global.cc (hunk 3 above)
Comment (spco) [inline source/global.cc]: output_parallel_statistics seems to have just been set by hand in the code, no associated parameter.

Comment (spco) [inline source/global.cc]: output_parallel_statistics seems to have just been set by hand in the code, no associated parameter.

**Comment by spco** [comment]
I agree - we might want to split into different stats files, but this will do for now. Most people might never use this postprocessor anyway, so it's not a big issue of long lines of data. Personally I rarely look at the statistics file on its own - easier to gnuplot than to find the correct column by hand!

I'll be making a matrix memory postprocessor, which might not be so appropriate to output to `statistics` but better as a separate file, as it might have a lot of columns (O(k^2) for a k-component system). Perhaps once we have a better way of matching data files in gnuplot we could think about how to split others?

**Comment by bangerth** [comment]
OK. Good plan forward. How about the test issue?

File: source/global.cc (hunk 3 above)
Comment (bangerth) [inline source/global.cc]: OK, thanks for checking.

**Comment by spco** [comment]
I guess a simple test could just be generating a 2x2 grid on a box, and a 4x4 (there might be such test already in existence), and compare the stats files? Memory consumption should be reproducible, right? Unless we might run into some issue of differences depending on the exact block allocation in memory, contiguous vs non-contiguous, say?

**Comment by spco** [comment]
Once this is done, do we need to include anything in `changes.h` either about the new postprocessor, or about the removal of `output_parallel_statistics` or `output_program_stats()`?

**Comment by bangerth** [comment]
Yes, any simple test with minimal run time will do. I suspect that things are going to be reproducible as long as you don't switch from 32-bit to 64-bit mode in ASPECT (which is not what we do for the tester).

changes.h: yes, please add an entry mentioning both the new postprocessor and the removal of the old flag.

**Comment by spco** [comment]
From my perspective, this is good to go once the tester is happy.

**Comment by bangerth** [comment]
OK to merge once the tester is happy.

**Comment by spco** [comment]
Conflicts resolved.

**Pull request merged**